txnsync/txngroups.go

```diff
@@ -58,3 +227,522 @@ func decodeTransactionGroups(bytes []byte) (txnGroups []transactions.SignedTxGro
+}
+
+func decodeTransactionGroups(bytes []byte) (txnGroups []transactions.SignedTxGroup, err error) {
+	if len(bytes) == 0 {
+		return nil, nil
+	}
+	var stub txGroupsEncodingStub
+	_, err = stub.UnmarshalMsg(bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	reconstructSignedTransactions(&stub)
+
+	txnGroups = make([]transactions.SignedTxGroup, stub.TransactionGroupCount)
+	for index, i := 0, 0; index < int(stub.TotalTransactionsCount); i++ {
+		size := 1
+		if i < len(stub.TransactionGroupSizes) {
+			size = int(stub.TransactionGroupSizes[i]) + 1
+		}
+		txnGroups[i].Transactions = stub.SignedTxns[index : index+size]
+		index += size
+	}
+
+	addGroupHashes(txnGroups)
+
+	return txnGroups, nil
+}
+
+func addGroupHashes(txnGroups []transactions.SignedTxGroup) {
+	for _, txns := range txnGroups {
+		if len(txns.Transactions) > 1 {
+			var txGroup transactions.TxGroup
+			txGroup.TxGroupHashes = make([]crypto.Digest, len(txns.Transactions))
+			for i, tx := range txns.Transactions {
+				txGroup.TxGroupHashes[i] = crypto.HashObj(tx.Txn)
+			}
+			groupHash := crypto.HashObj(txGroup)
+			for i := range txns.Transactions {
+				txns.Transactions[i].Txn.Group = groupHash
+			}
+		}
+	}
+}
+
+// deconstructs SignedTxn's into lists of fields and bitmasks
+func deconstructSignedTransactions(stub *txGroupsEncodingStub) {
+	bitmaskLen := (len(stub.SignedTxns)+7)/8 + 1
+	stub.BitmaskAuthAddr = make(bitmask, bitmaskLen)
+	stub.BitmaskLsig = make(bitmask, bitmaskLen)
+	stub.BitmaskMsig = make(bitmask, bitmaskLen)
+	stub.BitmaskSig = make(bitmask, bitmaskLen)
+	for i, txn := range stub.SignedTxns {
+		if !txn.Sig.MsgIsZero() {
+			stub.BitmaskSig.SetBit(i)
+			stub.Sig = append(stub.Sig, txn.Sig)
+			stub.SignedTxns[i].Sig = crypto.Signature{}
+		}
+		if !txn.Msig.MsgIsZero() {
+			stub.BitmaskMsig.SetBit(i)
+			stub.Msig = append(stub.Msig, txn.Msig)
+			stub.SignedTxns[i].Msig = crypto.MultisigSig{}
+		}
+		if !txn.Lsig.MsgIsZero() {
+			stub.BitmaskLsig.SetBit(i)
+			stub.Lsig = append(stub.Lsig, txn.Lsig)
+			stub.SignedTxns[i].Lsig = transactions.LogicSig{}
+		}
+		if !txn.AuthAddr.MsgIsZero() {
+			stub.BitmaskAuthAddr.SetBit(i)
+			stub.AuthAddr = append(stub.AuthAddr, txn.AuthAddr)
+			stub.SignedTxns[i].AuthAddr = basics.Address{}
+		}
+	}
+	stub.BitmaskAuthAddr = trimBitmask(stub.BitmaskAuthAddr)
+	stub.BitmaskLsig = trimBitmask(stub.BitmaskLsig)
+	stub.BitmaskMsig = trimBitmask(stub.BitmaskMsig)
+	stub.BitmaskSig = trimBitmask(stub.BitmaskSig)
+	deconstructTransactions(stub)
+}
+
+func deconstructTransactions(stub *txGroupsEncodingStub) {
+	bitmaskLen := (len(stub.SignedTxns)+7)/8 + 1
+	stub.BitmaskTxType = make(bitmask, bitmaskLen)
+	for i, txn := range stub.SignedTxns {
+		txTypeByte := protocol.TxTypeToByte(txn.Txn.Type)
+		if txTypeByte != 0 {
+			stub.BitmaskTxType.SetBit(i)
+			stub.TxType = append(stub.TxType, txTypeByte)
+		}
+		stub.SignedTxns[i].Txn.Type = ""
+	}
+	stub.BitmaskTxType = trimBitmask(stub.BitmaskTxType)
```

Comment:
That's interesting. So, you've picked a transaction type ( payment ), and optimized for that transaction type.
On the flip side, the rest of the transaction types are going to be stored quite inefficiently ( i.e. 8 bits/entry instead of 3 ).
Given that all transactions have a type, why don't we identify the most common transaction type, and optimize for that one ?

Reply:
something about the commas above is skewed. do you mean 124,745,162 ?

Reply:
Yup. 124 million, miscounted/typeod places above. A bit more than 10x the number of Payment txns.